File: src/packages/control-ui/src/api/ppp_tls_tunnel_codec.ts
export class PppTlsTunnelCodec {
  public enableHdlc: boolean;
  constructor(enableHdlc: boolean = true) {
    this.enableHdlc = enableHdlc;
  }

  encodeFrame(proto: number, payload: Uint8Array): Uint8Array {
    const hdlcLen = this.enableHdlc ? 2 : 0;
    const bodyLen = hdlcLen + 2 + payload.length;
    const frame = new Uint8Array(2 + bodyLen);
    const view = new DataView(frame.buffer, frame.byteOffset, frame.byteLength);

    view.setUint16(0, bodyLen);
    let offset = 2;
    if (this.enableHdlc) {
      frame[offset] = 0xff;
      frame[offset + 1] = 0x03;
      offset += 2;
    }

    view.setUint16(offset, proto);
    offset += 2;
    frame.set(payload, offset);
    return frame;
  }

  decodeFrame(data: Uint8Array): { proto: number; payload: Uint8Array } | null {
    if (data.length < 2) return null;
    const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
    const bodyLen = view.getUint16(0);
    if (data.length < 2 + bodyLen) return null;

    let offset = 2;
    if (this.enableHdlc) {
      if (bodyLen < 2) return null;
      if (data[offset] !== 0xff || data[offset + 1] !== 0x03) return null;
      offset += 2;
    }

    if (bodyLen < (this.enableHdlc ? 4 : 2)) return null;
    const proto = view.getUint16(offset);
    offset += 2;

    const payloadLen = 2 + bodyLen - offset;
    const payload = data.slice(offset, offset + payloadLen);
    return { proto, payload };
  }
}

File: labs/mobile/donor-archive/tunnel/PppTlsTunnelCodec.kt
package com.luminet.android.tunnel

import java.nio.ByteBuffer

class PppTlsTunnelCodec(val enableHdlc: Boolean = true) {

    fun encodeFrame(proto: Short, payload: ByteArray): ByteArray {
        val hdlcLen = if (enableHdlc) 2 else 0
        val bodyLen = hdlcLen + 2 + payload.size
        val buf = ByteBuffer.allocate(2 + bodyLen)

        buf.putShort(bodyLen.toShort())
        if (enableHdlc) {
            buf.put(0xFF.toByte())
            buf.put(0x03.toByte())
        }
        buf.putShort(proto)
        buf.put(payload)
        return buf.array()
    }

    fun decodeFrame(data: ByteArray): Pair<Short, ByteArray>? {
        if (data.size < 2) return null
        val buf = ByteBuffer.wrap(data)
        val bodyLen = buf.getShort().toInt() and 0xFFFF
        if (data.size < 2 + bodyLen) return null

        var offset = 0
        if (enableHdlc) {
            if (bodyLen < 2) return null
            val b0 = buf.get().toInt() and 0xFF
            val b1 = buf.get().toInt() and 0xFF
            if (b0 != 0xFF || b1 != 0x03) return null
            offset += 2
        }

        if (bodyLen < offset + 2) return null
        val proto = buf.getShort()
        val payloadLen = bodyLen - offset - 2
        val payload = ByteArray(payloadLen)
        buf.get(payload)
        return Pair(proto, payload)
    }
}
